singularity-eos/base/error_utils.hpp

```diff
@@ -0,0 +1,88 @@
+//------------------------------------------------------------------------------
+// © 2021-2024. Triad National Security, LLC. All rights reserved.  This
+// program was produced under U.S. Government contract 89233218CNA000001
+// for Los Alamos National Laboratory (LANL), which is operated by Triad
+// National Security, LLC for the U.S.  Department of Energy/National
+// Nuclear Security Administration. All rights in the program are
+// reserved by Triad National Security, LLC, and the U.S. Department of
+// Energy/National Nuclear Security Administration. The Government is
+// granted for itself and others acting on its behalf a nonexclusive,
+// paid-up, irrevocable worldwide license in this material to reproduce,
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#ifndef SINGULARITY_EOS_BASE_ERROR_UTILS_HPP_
+#define SINGULARITY_EOS_BASE_ERROR_UTILS_HPP_
+
+#include <cmath>
+#include <type_traits>
+
+#include <ports-of-call/portability.hpp>
+
+namespace singularity {
+namespace error_utils {
+
+using PortsOfCall::printf;
+
+constexpr double _NORMAL_FACTOR = 1.0e10;
+
+struct is_normal_or_zero {
+  template <typename valT>
+  bool PORTABLE_FORCEINLINE_FUNCTION operator()(valT value) const {
+    static_assert(std::is_floating_point_v<valT>);
+    return (value == valT{0}) ||
+           (std::isnormal(_NORMAL_FACTOR * value) && std::isnormal(value));
+  }
+};
```

Comment:
These as structs is fine. But why structs instead of just functions?

Reply:
Stylistically I like functors better than functions because I find it easier to reason about the type. I could have used lambdas here, but I find structs with templates more readable even if they're a bit more verbose.

I also thought it would be more extensible to be explicit that zero be the same type as the template parameter rather than letting the compiler compare to an `int` literal or maybe interpreting zero as a different type.

Reply:
The type of zero thing isn't really that important, but you might imagine a future at some point where the types of the values are unit aware somehow and this explicitness might be favorable.

If you'd like, I can wrap these in an `impl` namespace to signal that they're implementation details.

Reply:
Actually, these might be useful on their own. I'll leave them as-is if that's okay.

Reply:
Yeah it's fine to leave them.